### PR TITLE
Better support for initializing without explicit initializer(issue#4419)

### PR DESCRIPTION
**Pull request by DjangoPeng**
When using `tf.get_variable(name='foo', shape=(42,), dtype=tf.int32)` without explicit initializer, it causes the error:

```
TypeError: Expected int32, got -1.7320508075688772 of type 'float' instead.
```

The reason is as follows:
If no initializer provided, it would use the default initializer `uniform_unit_scaling_initializer` which just fit with `float`(notice that sqrt(3)==1.7320...). While, it always conflict with the required integer dtype.

The current solution is initialize the integer by `init_ops.zeros_initializer` when no initializer provided.

reference the [issus#4419](https://github.com/tensorflow/tensorflow/issues/4419)

### CODE REVIEW

**Comment by ebrevdo** [comment]
Perhaps we should just throw a cleaner error?  Initializer should be explicit imo

**Comment by ebrevdo** [comment]
Nevermind. That's not backwards compatible.

**Comment by DjangoPeng** [comment]
@ebrevdo But I think Tensorflow should provide the default initializer for integer, such as float. And in many platform and language, they have the build-in mechanism to handle the case

File: tensorflow/python/ops/variable_scope.py
Comment (rohan100jain) [inline tensorflow/python/ops/variable_scope.py]: Wouldn't it be better for this to be a function instead? In that case, for partitioned variables and such, there would be better support.

**Review by rohan100jain — CHANGES_REQUESTED** [review]
Can you also add test cases in variable_scope_test.py to make sure that the issue is fixed?

**Comment by DjangoPeng** [comment]
@rohan100jain Where should I define the function to handle the issue, init_op? or variable_scope?

**Comment by DjangoPeng** [comment]
@rohan100jain  Please review

**Comment by DjangoPeng** [comment]
@lukaszkaiser @rohan100jain @ebrevdo, @theweiho @keveman  Can one of you review the PR?

File: tensorflow/python/framework/dtypes.py
Comment (ebrevdo) [inline tensorflow/python/framework/dtypes.py]: can you avoid moving around methods?  revert this part of your change?

File: tensorflow/python/ops/array_ops.py
Comment (ebrevdo) [inline tensorflow/python/ops/array_ops.py]: i dont think we need a dtype argument.

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: why do complex values get zeros initialized?  i dont think it'll even reach this point since complex types are is_floating.

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: can we punt on string initialization till another PR?  i'm not in love with the tf.strings method and would prefer you just leave that out of this one.  keep it focused.

File: tensorflow/python/ops/variable_scope_test.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope_test.py]: there is a variable scope test in kernel_tests; please put your tests there following the existing style of the file.

Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope_test.py]: (that includes using indentation and whitespace consistent with how we do it everywhere else)

**Review by ebrevdo — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Yep, complex is not tf.float, so I used `dtype.is_complex`, and the argument passed to `zeros_initializer` is `dtype.base_type=tf.complex64/128`. At the same time, initialize complex by zeros is reasonable in general.

File: tensorflow/python/ops/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope_test.py]: @rohan100jain So I don't need to add the test cases in `variable_scope_test.py`?

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: I think tf.string is the same issue regarding initialization, so put them in one PR. If you insist that they are different, I can split them to reopen a PR for tf.string

File: tensorflow/python/ops/array_ops.py
Comment (DjangoPeng) [inline tensorflow/python/ops/array_ops.py]: Yep, you're right~ This function is just for `tf.string`, so it doesn't need a dtype argument

**Comment by ebrevdo** [comment]
Please send a separate PR for tf.string, I think I'm not the best person to
review that (maybe andrew selle)

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: i'm not convinced that initializing complex with zeros is reasonable in general.  do you have some references showing neural networks that use complex parameters and suggest that initializing them with zeros is the right way to do it?

certainly having this be an undocumented feature is problematic; and since complex are complex floating point, it also seems very reasonable to initialize them like regular floating point.

File: tensorflow/python/ops/variable_scope_test.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope_test.py]: you **need** to add test cases; but put them in kernel_tests/variable_scope_test; not in this new file.

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Yep, zeros initialization may not be the most appropriate if you just focus on NN. But I always    think tensorflow as a general Machine Learning platform, not only for NN, but also for other cases and solutions. As we can see the updated tutorials for [Non-ML Applications](https://www.tensorflow.org/versions/r0.11/tutorials/mandelbrot/index.html#mandelbrot-set), so I prefer to keep the initializer for complex.

Regarding zeros or sqrt(3), I think both are ok. Do you think so?

File: tensorflow/python/ops/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope_test.py]: @ebrevdo Gotcha~I'm on it.

File: tensorflow/python/framework/dtypes.py
Comment (DjangoPeng) [inline tensorflow/python/framework/dtypes.py]: Fine, I do the adjustment cause there are two assigners told me this is a custom of Google, and let me fix it in the previous PRs. So, this time, I adjusted them in passing, while it looks like it's not fit for everyone. I'll revert that~

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: s/variable ./variable /

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: Raise a ValueError, and instead of dtype provide the name of the variable being initialized.  Otherwise it's not a very helpful error string.

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: since the complex case is ambiguous enough for us to have a discussion about what value it should be, let's just fall back on raising an error in the complex case.

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: add spaces around the equality sign.

**Review by ebrevdo — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Yep, so we can handle the complex dtype here as well

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Fixed

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Raise ValueError, but I really don't know how to give an appropriate error string. 
:(

Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: Can't you pass name directly through to this function so you can use it in the error message?

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: 'x%d' % i

here and elsewhere

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: don't use _get_variable_list; which is private!

use list(x)

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: Replace private function `_get_variable_list()` with `list()`

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: Set string format to `'x%d' % i`

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: Because using `axis0_into2_partitioner`,  `y` is required a replica shape in 1-D of `x`.

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: Delete redundant spaces here and elsewhere.

**Comment by DjangoPeng** [comment]
@ebrevdo Please review.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by DjangoPeng** [comment]
@drpngx @ebrevdo It seems like everything goes well. ⊙▽⊙

**Comment by drpngx** [comment]
Did you address all changes?

**Comment by DjangoPeng** [comment]
yep，including necessary comments

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: capital letters only allowed for module-level static constants.  this should be just 'types'.

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: for (i, dtype) in enumerate(types):

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: i can't tell but is this within the 80 character limit?

**Comment by ebrevdo** [comment]
More comments.

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: Fix capital here and eleswhere.

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Fix typesetting for 80 chars limit here and elsewhere.

**Comment by DjangoPeng** [comment]
@ebrevdo Could you review and merge this PR, please?

**Comment by DjangoPeng** [comment]
@ebrevdo Would you mind testing the PR?

**Comment by drpngx** [comment]
Jenkins, test this please.

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: whitespace: (3,4) should be (3,<space>4)

here and everywhere else.

**Review by ebrevdo — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/variable_scope_test.py
Comment (DjangoPeng) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: fixed typesetting here and elsewhere

**Comment by DjangoPeng** [comment]
@ebrevdo Please check.

**Comment by DjangoPeng** [comment]
@ebrevdo @drpngx  I revert all commits before and re-push the branch with more informative comments. Now, it would be better to review and merge. 
## Change log:
### dtypes.py
Add `is_bool()`  function for `if` judgement in `_get_default_initializer`. And `DT_BOOL` can use `zeros_initializer` to initialize with `FALSE`, so I combine it with `DT_INT` and `DT_UINT` to avoid duplicating code block.

### variable_scope.py
Add `_get_default_initializer()` private function to provide a default initializer and a corresponding value when no explicit initializer provided. Now, it can handle `DT_FLOAT/DT_INT/DT_UINT/DT_BOOL`. In other dtype case, it would raise a `ValueError` with info `\"An initializer for variable %s is required\" % name`.

### variable_scope_test.py
Add `testInitFromNonInitializer()` for single_variable and partitioned_variable.

File: tensorflow/python/ops/variable_scope.py
Comment (vrv) [inline tensorflow/python/ops/variable_scope.py]: End one-line docstrings with a period

File: tensorflow/python/ops/variable_scope.py
Comment (vrv) [inline tensorflow/python/ops/variable_scope.py]: maybe print the type name so they know why it doesn't have a default initializer

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Give a period in end-line.

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: Give dtype name in `ValueError` for more info to debug.

**Comment by DjangoPeng** [comment]
@vrv Please review.

**Comment by vrv** [comment]
Running a test but going to let @ebrevdo and @rohan100jain give the final approval.

@tensorflow-jenkins test this please

**Comment by DjangoPeng** [comment]
@vrv It seems like there are some internal errors?

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: I found the reason of testing failure. Fixed.

**Review by DjangoPeng — COMMENTED** [review]
@vrv Please test again, thanks!

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@rohan100jain @ebrevdo does this look good?  I ca merge if so.

File: tensorflow/python/ops/variable_scope.py
Comment (rohan100jain) [inline tensorflow/python/ops/variable_scope.py]: I think initializing_from_value isn't documented in get_variable. So perhaps clarify that here?

File: tensorflow/python/ops/variable_scope.py
Comment (rohan100jain) [inline tensorflow/python/ops/variable_scope.py]: vrv@, ebrevdo@ - does it make a difference whether this is a function or a Tensor? Or performance / memory wise its all the same?

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: yep, `initializing_from_value` isn't documented in `get_variable`. While, `get_variable` calls `_get_single_variable` and `_get_partitioned_variable`, and both them define it. From my perspective, the comment is a little unclear, but I think the name itself is easy to understand. Do we really need to clarify here? instead of above?

File: tensorflow/python/ops/variable_scope.py
Comment (DjangoPeng) [inline tensorflow/python/ops/variable_scope.py]: If we want to use a function to handle this case, we have to define a set of functions to cooperate with it. While it would only run once and then copy value to the variable, which is not expensive. Do you think so?

**Comment by DjangoPeng** [comment]
@ebrevdo @rohan100jain @vrv @drpngx Can one of you verify the patch?

**Comment by danmane** [comment]
@DjangoPeng What's the status here - should I poke rohan/ebrevdo to re-review?
Also, it now has conflicts, so you will need to rebase.

**Comment by DjangoPeng** [comment]
@danmane I think it's ready to merge, and I will rebase it after work today. I hope them can accept the PR, which is helpful from my perspective.

File: tensorflow/python/ops/variable_scope.py
Comment (ebrevdo) [inline tensorflow/python/ops/variable_scope.py]: i believe zeros_initializer was recently converted to a function instead of a tensor.

**Comment by ebrevdo** [comment]
once you merge we can run a test.

**Review by ebrevdo — APPROVED** [review]

**Comment by DjangoPeng** [comment]
@danmane @ebrevdo I have solved the conflict and update `tf.global_variables_initializer()` API. Please check and test.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by DjangoPeng** [comment]
@ebrevdo @drpngx @danmane It looks everything goes well, would you mind merging it?

**Pull request merged**
